frontend/catalyst/jax_primitives.py

```diff
@@ -551,13 +559,87 @@ def _apply_registered_pass_lowering(
 
 
 #
-# func
+# module
 #
-mlir_fn_cache: Dict["catalyst.jax_tracer.Function", Any] = {}
+
+mlir_fn_cache: Dict[Any, Any] = {}
+
+
+@module_p.def_impl
+def _module_def_impl(*args, call_jaxpr, fn):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _module_to_mlir(ctx, *args, call_jaxpr, fn, call=True):
+
+    if not isinstance(fn, qml.QNode) and not call:
+        _func_lowering(ctx, *args, call_jaxpr=call_jaxpr, fn=fn, call=call)
+        return None
+
+    output_types = list(map(mlir.aval_to_ir_types, ctx.avals_out))
+    flat_output_types = util.flatten(output_types)
+
+    if mlir_fn_cache.get(fn) and not call:
+        return None
+
+    symbol_table = ctx.module_context.symbol_table
```

Comment:
What about 
```python
if mlir_fn_cache.get(fn):
   if not call:
      return None
   else:
      func_op = mlir_fn_cache[fn]
      ...
      call = ...
      return call.results

else:
   func_op = ...create new qnode module func op...
   if not call:
      ...
   else:
      ...
```

Personally I think this structure is clearer

Reply:
A lot has been rewritten, please take a look. There is no longer a `call=` parameter.